# Script Contents
### -Import Libraries

### -Import Data Part 1

### -Data Wrangling

### -Data Dictionary

### -Sub

### -Confirm and Investigate Combined Data

### -Export Combined Data

# .01 Importing Libraries

In [4]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# .02 Importing Data

In [8]:
# Creating a path for my main project folder
path = r"C:\Users\jacks\Instacart Basket Analysis 081424"

In [10]:
# Import the order data, product data, and department data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# .03 Data Wrangling

In [11]:
# Dropping eval_set column from the orders dataframe
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [13]:
# Setting the df_ords variable to equal the orders dataframe with dropped eval_set column
df_ords = df_ords.drop(columns = ['eval_set'])

In [17]:
# Count the number of values for each value in the 'order' dataframe 'days_since_prior_order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [19]:
# Rename the 'order_dow' column in the 'orders' dataframe
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [21]:
# Check the orders dataframe to see if the column rename was successful 
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [23]:
# Change the data type of the 'order_id' column in the orders dataframe
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [25]:
# Check the data type of the 'order_id' column
df_ords['order_id'].dtype

dtype('O')

In [27]:
# Check out the department dataframe
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [29]:
# Transpose the department dataframe
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [31]:
# Set variable for transposed department dataframe
df_dep_t = df_dep.T

In [33]:
# Check out the df_dep_t variable
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [35]:
# Add an index to the department dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [39]:
# Take the first row of df_dep_t for the header
new_header = df_dep_t.iloc[0]

In [41]:
# Call on the new_header variable to check it out
new_header

0    department
Name: department_id, dtype: object

In [43]:
# Take the data under the header row for a new df and check out the results
df_dep_t_new = df_dep_t[1:]
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [45]:
# set the new header variable as the df header and check out the results
df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# .04 Data Dictionary

In [47]:
# Turn the 'df_dep_t_new' dataframe into a dictionary and check out the results
data_dict = df_dep_t_new.to_dict('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [49]:
# Remind myself of the first few rows of df_prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [51]:
# Use data dictionary to find out what '19' in the department_id column of the df_prods dataframe stands for
print(data_dict.get('19'))

{'department': 'snacks'}


# .05 Subsetting

In [57]:
# Look for every row in the df_prad's department_id column where there's a value of '19'
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [61]:
# Look for every row in the df_prod dataframe where there's a value of 19 in the department_id column
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [65]:
# Create a new datafram, df_snacks, that only contains rows from the df_prod dataframe where '19' is the value in the department_id column
# Check out the results
df_snacks = df_prods[df_prods['department_id']==19]
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [71]:
# Alternatives to creating a snacks subset of the df_prods dataframe that use the 'loc' and 'isin()' functions
df_snacks_2 = df_prods.loc[df_prods['department_id']==19]
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

# Task 4.4

## 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [75]:
# Remind myself of the df_ords variables
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [77]:
# Change user_id variable in df_ords to string data type since it is an identifier variable
df_ords['user_id'] = df_ords['user_id'].astype('str')
# Check the results
df_ords['user_id'].dtype

dtype('O')

## 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [12]:
# Rename the 'order_hour_of_day' column to 'hour_of_order' column on a new df 'df_ords_2'
df_ords_2 = df_ords.rename(columns = {'order_hour_of_day' : 'hour_of_order'}, inplace = False)

In [14]:
# Check the results of the rename
df_ords_2.head()

order_id  user_id eval_set  order_number  order_dow  hour_of_order  \
0   2539329        1    prior             1          2              8   
1   2398795        1    prior             2          3              7   
2    473747        1    prior             3          3             12   
3   2254736        1    prior             4          4              7   
4    431534        1    prior             5          4             15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [16]:
# The rename appears to have changed the original df_ords dataframe
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [123]:
# I am not sure why the rename to a new df, 'df_ords_2' isn't working
# For now, I am reverting df_ords back to the original column names as to not overwrite the dataframe
df_ords.rename(columns = {'hour_of_order' : 'order_hour_of_day'}, inplace = True)

In [125]:
# Check the results of the rename
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [127]:
# Count the values in the 'order_hour_of_day' column
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### 10am appears to be the busiest hour

## 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [131]:
# I am usiing the data dictionary set up earlier in the exercise
print(data_dict.get('4'))

{'department': 'produce'}


### 4 in the df_prod's department_id column refers to the produce department

## 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [152]:
# I need a reminder of the numeric value 'breakfast' corresponds with inside the dp_dep_t_new dataframe
print(df_dep_t_new[df_dep_t_new['department'] == 'breakfast'])

department_id department
14             breakfast


In [154]:
# Create a subset of the df_prods dataframe that only contains breakfast data
df_breakfast = df_prods[df_prods['department_id']==14]
# Check out the results
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [161]:
# I need a reminder of the df_dep_t_new values for alcohol, deli, beverages, and meat/seafood
print(df_dep_t_new[df_dep_t_new['department'].isin(['alcohol','deli','beverages','meat seafood'])])

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli


In [163]:
# Create a subset of the df_prods dataframe containing all departments related to dinner parties
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]
# Check out the results
df_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [168]:
# Check some basic info for the df_dinner_party dataframe
df_dinner_party.shape

(7650, 5)

### There are 7,650 rows in the df_dinner_party dataframe

## 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [182]:
# Pull data for user_id '1' from df_ords
df_ords[df_ords['user_id']=='1']

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### Since there are only 11 rows of data for user_id '1', this is all the info I think I can extract. The next question asks about analysis so I will wait to further investigate this data.

##  10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [197]:
# Describe user_id '1' data
df_ords[df_ords['user_id']=='1'].describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Basic stats are listed. Nothing strands out as particularly strange to me. 

## 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [201]:
# Exporting to Prepared Data folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

## 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [204]:
# Exporting to Prepared Data folder
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))